src/application-manager.coffee

```diff
@@ -0,0 +1,875 @@
+
+		_.forEach networks, (network) ->
+			appId = network.appId
+			apps[appId] ?= { appId }
+			apps[appId].networks ?= {}
+			apps[appId].networks[network.name] = network.config
+
+		_.forEach volumes, (volume) ->
+			appId = volume.appId
+			apps[appId] ?= { appId }
+			apps[appId].volumes ?= {}
+			apps[appId].volumes[volume.name] = volume.config
+
+		# We return the apps as an array
+		return _.values(apps)
+
+	getCurrentForComparison: =>
+		Promise.join(
+			@services.getAll()
+			@networks.getAll()
+			@volumes.getAll()
+			(services, networks, volumes) =>
+				return @_buildApps(services, networks, volumes)
+		)
+
+	getCurrentApp: (appId) =>
+		Promise.join(
+			@services.getAllByAppId(appId)
+			@networks.getAllByAppId(appId)
+			@volumes.getAllByAppId(appId)
+			(services, networks, volumes) =>
+				return @_buildApps(services, networks, volumes)[0]
+		)
+
+	getTargetApp: (appId) =>
+		@db.models('app').where({ appId }).select()
+		.then ([ app ]) =>
+			return if !app?
+			@normaliseAndExtendAppFromDB(app)
+
+	# Compares current and target services and returns a list of service pairs to be updated/removed/installed.
+	# The returned list is an array of objects where the "current" and "target" properties define the update pair, and either can be null
+	# (in the case of an install or removal).
+	compareServicesForUpdate: (currentServices, targetServices) ->
+		Promise.try ->
+			removePairs = []
+			installPairs = []
+			updatePairs = []
+			targetServiceIds = _.map(targetServices, 'serviceId')
+			currentServiceIds = _.uniq(_.map(currentServices, 'serviceId'))
+
+			toBeRemoved = _.difference(currentServiceIds, targetServiceIds)
+			_.forEach toBeRemoved, (serviceId) ->
+				servicesToRemove = _.filter(currentServices, (s) -> s.serviceId == serviceId)
+				_.map servicesToRemove, (service) ->
+					removePairs.push({
+						current: service
+						target: null
+						serviceId
+					})
+
+			toBeInstalled = _.difference(targetServiceIds, currentServiceIds)
+			_.forEach toBeInstalled, (serviceId) ->
+				servicesToInstall = _.filter(targetServices, (s) -> s.serviceId == serviceId)
+				_.map servicesToInstall, (service) ->
+					installPairs.push({
+						current: null
+						target: service
+						serviceId
+					})
+
+			toBeMaybeUpdated = _.intersection(targetServiceIds, currentServiceIds)
+			currentServicesPerId = {}
+			targetServicesPerId = {}
+			_.forEach toBeMaybeUpdated, (serviceId) ->
+				currentServiceContainers = _.filter currentServices, (service) ->
+					return service.serviceId == serviceId
+				targetServicesForId = _.filter targetServices, (service) ->
+					return service.serviceId == serviceId
+				throw new Error("Target state includes multiple services with serviceId #{serviceId}") if targetServicesForId.length > 1
+				targetServicesPerId[serviceId] = targetServicesForId[0]
+				if currentServiceContainers.length > 1
+					currentServicesPerId[serviceId] = _.maxBy(currentServiceContainers, 'createdAt')
```

Comment:
How can this happen? Also could it choose the wrong one due to a combination of NTP and reboots causing the timestamp of the new one to be less than the old one?

Reply:
This can happen in rare scenarios, like an unexpected poweroff while doing a handover. I figured choosing the latest should be a reasonable best effort - both this case being hit *and* an ntp issue at the same time seems rare enough not to care?